FirebaseMessaging/Sources/Token/FIRMessagingTokenManager.m

```diff
@@ -655,42 +653,51 @@ - (void)setAPNSToken:(NSData *)APNSToken withUserInfo:(NSDictionary *)userInfo {
-                                    @"apnsTokenData was set to nil during token reset. Aborting");
-            return;
-          }
+    [self.installations installationIDWithCompletion:^(NSString *_Nullable identifier,
+                                                       NSError *_Nullable error) {
+      FIRMessaging_STRONGIFY(self);
+      if (self == nil) {
+        FIRMessagingLoggerError(kFIRMessagingMessageCodeInstanceID017,
+                                @"Instance ID shut down during token reset. Aborting");
+        return;
+      }
+      if (self.currentAPNSInfo == nil) {
+        FIRMessagingLoggerError(kFIRMessagingMessageCodeInstanceID018,
+                                @"apnsTokenData was set to nil during token reset. Aborting");
+        return;
+      }
 
-          NSMutableDictionary *tokenOptions = [@{
-            kFIRMessagingTokenOptionsAPNSKey : self.currentAPNSInfo.deviceToken,
-            kFIRMessagingTokenOptionsAPNSIsSandboxKey : @(isSandboxApp)
-          } mutableCopy];
-          if (self.firebaseAppID) {
-            tokenOptions[kFIRMessagingTokenOptionsFirebaseAppIDKey] = self.firebaseAppID;
-          }
+      NSMutableDictionary *tokenOptions = [@{
+        kFIRMessagingTokenOptionsAPNSKey : self.currentAPNSInfo.deviceToken,
+        kFIRMessagingTokenOptionsAPNSIsSandboxKey : @(isSandboxApp)
+      } mutableCopy];
+      if (self.firebaseAppID) {
+        tokenOptions[kFIRMessagingTokenOptionsFirebaseAppIDKey] = self.firebaseAppID;
+      }
 
-          for (FIRMessagingTokenInfo *tokenInfo in invalidatedTokens) {
-            [self fetchNewTokenWithAuthorizedEntity:tokenInfo.authorizedEntity
-                                              scope:tokenInfo.scope
-                                         instanceID:identifier
-                                            options:tokenOptions
-                                            handler:^(NSString *_Nullable token,
-                                                      NSError *_Nullable error){
+      for (FIRMessagingTokenInfo *tokenInfo in invalidatedTokens) {
+        [self fetchNewTokenWithAuthorizedEntity:tokenInfo.authorizedEntity
+                                          scope:tokenInfo.scope
+                                     instanceID:identifier
+                                        options:tokenOptions
+                                        handler:^(NSString *_Nullable token,
+                                                  NSError *_Nullable error){
 
-                                            }];
-          }
-        }];
+                                        }];
+      }
+      if ([self->_tokenStore cachedTokenInfos].count == 0) {
+        [self tokenWithAuthorizedEntity:self.fcmSenderID
+                                  scope:kFIRMessagingDefaultTokenScope
+                                options:tokenOptions
+                                handler:^(NSString *_Nullable FCMToken, NSError *_Nullable error){
+
```

Comment:
Do we need to do anything in this handler? If not, can we add a comment saying why we aren't doing anything?

Reply:
I can add a comment here.